File: src/leetcode/python/medium/CourseScheduleII_210.py
class CourseScheduleII_210(object):
    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """
        def dfs(course):
            if visited[course] == 0:
                visited[course] = 1
                onstack[course] = 1
                if course in neigh:
                    for j in neigh[course]:
                        if not dfs(j):
                            return False
                topo.append(course)
                onstack[course] = 0
            elif onstack[course] == 1:
                return False
            return True

        if len(prerequisites) == 0:
            return [i for i in range(numCourses)]
        neigh = {}
        for i, j in prerequisites:
            if j not in neigh:
                neigh[j] = set()
            neigh[j].add(i)
        visited, onstack = [0] * numCourses, [0] * numCourses
        topo = []
        for i in range(numCourses):
            if not dfs(i):
                return []
        return topo[::-1]

File: src/leetcode/java/CourseScheduleII_210.java
package leetcode.java.medium;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;

public class CourseScheduleII_210 {
  public int[] findOrder(int numCourses, int[][] prerequisites) {
    int[] order = new int[numCourses];
    for (int i = 0; i < numCourses; i++)
      order[i] = i;
    if (prerequisites.length == 0) return order;

    int[] visited = new int[numCourses];
    int[] onStack = new int[numCourses];
    HashMap<Integer, HashSet<Integer>> neigh = new HashMap<>();
    for (int i = 0; i < prerequisites.length; i++) {
      neigh.putIfAbsent(prerequisites[i][1], new HashSet<>());
      neigh.get(prerequisites[i][1]).add(prerequisites[i][0]);
    }
    List<Integer> topo = new ArrayList<>();

    for (int i = 0; i < numCourses; i++) {
      if (!dfs(i, prerequisites, visited, onStack, topo, neigh)) return new int[0];
    }

    for (int i = topo.size() - 1, k = 0; i >= 0; i--, k++) {
      order[k] = topo.get(i);
    }
    return order;
  }

  private boolean dfs(int i, int[][] prerequisites, int[] visited, int[] onStack,
      List<Integer> topo, HashMap<Integer, HashSet<Integer>> neigh) {
    if (visited[i] == 0) {
      visited[i] = 1;
      onStack[i] = 1;
      if (neigh.containsKey(i)) {
        for (int n : neigh.get(i)) {
          if (!dfs(n, prerequisites, visited, onStack, topo, neigh)) return false;
        }
      }
      topo.add(i);
      onStack[i] = 0;
    } else if (onStack[i] == 1) return false;
    return true;
  }
}
